walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,150 @@
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+    Collection<String> allAttendees = new ArrayList<>();
+    allAttendees.addAll(requestAttendees);
+    allAttendees.addAll(optionalAttendees);
+    long requestDuration = request.getDuration();
+    int numRequestSlots = (int) requestDuration / 30;
+    int numEvents = events.size();
+
+    boolean[] mandatoryTimeSlots = new boolean[48];
+    boolean[] allTimeSlots = new boolean[48];
+
+    // if the duration is longer than a day there are no options
+    if (requestDuration >= TimeRange.WHOLE_DAY.duration() + 1)
+      return Arrays.asList();
+    // if there are no attendees, the time range is the whole day
+    if (events.size() == 0 || allAttendees.size() == 0) 
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    
+    // sets up timeSlots array; anywhere left true has all the requestAttendees available
+    for (Event event : events) { 
+      mandatoryTimeSlots = convertToTimeSlots(requestAttendees, events);
+      allTimeSlots = convertToTimeSlots(allAttendees, events);    
+    }
+
+    ArrayList<Integer> possibleSlots = convertToSlots(mandatoryTimeSlots, numRequestSlots);
+    ArrayList<Integer> allSlots = convertToSlots(allTimeSlots, numRequestSlots);
+    
+    ArrayList<TimeRange> possibleTimes = convertToTimes(possibleSlots, numRequestSlots, numEvents);
+    ArrayList<TimeRange> allTimes = convertToTimes(allSlots, numRequestSlots, numEvents);
+
+    if (allTimes.isEmpty() && !possibleTimes.contains(TimeRange.WHOLE_DAY))
+        return possibleTimes;
+    return allTimes;
+  }
+  
+  public static boolean[] convertToTimeSlots (Collection<String> requestAttendees, Collection<Event> events) {
+    boolean[] timeSlots = new boolean[48];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+    int numEvents = events.size();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        numEvents--;
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) 
+              timeSlots[start + i] = false;
+            break;
+          }
+        }
+        if(numEvents == 0)
+            break;
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < 48; i++) {
```

Comment:
Can you make 48 a constant and replace its usage here and other places (since you use it a couple times, using a constant just makes it obvious what it means)